entity-framework/core/querying/null-semantics.md

```diff
@@ -0,0 +1,131 @@
+description: Information on how Entity Framework Core handles null comparisons in queries 
+author: maumar
+ms.date: 11/11/2020
+uid: core/querying/null-semantics
+---
+# Query null semantics
+
+## Introduction
+
+SQL databases operate on 3-valued logic (`true`, `false`, `null`) when performing comparisons, as opposed to boolean logic of C#. When translating LINQ queries to SQL, EF Core tries to compensate for the difference by introducing additional null checks for some elements of the query.
+To illustrate this, lets define the following entity:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsEntity.cs#Entity)]
+
+and issue several queries:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#BasicExamples)]
+
+First two queries produce simple comparison. In the first query, both columns are non-nullable so null checks are not needed. In the second query, `NullableInt` could contain `null`, however `Id` is non-nullable. Comparing `null` to non-null yields `null` as a result, which would be filtered out by `WHERE` operation. No additional terms are needed either.
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[Id] = [e].[Int]
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[Id] = [e].[NullableInt]
+```
+
+Third query introduces the null check. When `NullableInt` is `null` the comparison `Id <> NullableInt` yields `null`, which would be filtered out by `WHERE` operation. However from the boolean logic perspective this case should be returned as part of the result. Hence EF Core adds the necessary check to ensure that.
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ([e].[Id] <> [e].[NullableInt]) OR [e].[NullableInt] IS NULL
+```
+
+Queries four and five show the pattern when both columns are nullable. It's worth noting that `<>` operation produces more complicated (and slower) query than equality operation.
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ([e].[String1] = [e].[String2]) OR ([e].[String1] IS NULL AND [e].[String2] IS NULL)
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE (([e].[String1] <> [e].[String2]) OR ([e].[String1] IS NULL OR [e].[String2] IS NULL)) AND ([e].[String1] IS NOT NULL OR [e].[String2] IS NOT NULL)
+```
+
+## Null semantics in functions
+
+Many functions in SQL can only return `null` result if some of their arguments are `null`. EF Core takes advantage of this to produce more efficient queries.
+The query below illustrates the optimization:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#Functions)]
+
+The generated SQL is as follows (we don't need to evaluate the `SUBSTRING` function):
```

Comment:
```suggestion
The generated SQL is as follows (we don't need to evaluate the `SUBSTRING` function since it will be only null when either of the arguments to it is null.):
```